examples/vision/pointnet.py

```diff
@@ -0,0 +1,295 @@
+            test_pts.append(trimesh.load(f).sample(n_pts))
+            test_labels.append(i)
+
+    return (
+        np.array(train_pts),
+        np.array(test_pts),
+        np.array(train_labels),
+        np.array(test_labels),
+    )
+
+
+"""
+Set the number of points to sample and batch size and parse the dataset. This can take
+~5minutes to complete.
+"""
+
+N_PTS = 2048
+NUM_CLASSES = 10
+BATCH_SIZE = 32
+
+X_train, X_test, y_train, y_test = parse_dataset(N_PTS)
+
+"""
+Our data can now be read into a `tf.data.Dataset()` object. We set the shuffle buffer
+size to the entire size of the dataset as prior to this the data is ordered by class.
+Data augmentation is important when working with point cloud data. We create a
+augmentation function to jitter, shuffle and rotate the train dataset.
+"""
+
+
+def augment(pts, label):
+
+    # jitter points
+    pts += tf.random.uniform(pts.shape, -0.005, 0.005, dtype=tf.float64)
+    # shuffle points
+    pts = tf.random.shuffle(pts)
+    # random rotation about the z axis
+    theta = np.random.uniform(-np.pi, np.pi)
+    c = np.cos(theta)
+    s = np.sin(theta)
+    R = tf.constant([[c, -s, 0.0], [s, c, 0.0], [0.0, 0.0, 1.0]], tf.float64)
+    pts = tf.einsum("...ij,...kj->...ki", R, pts)
+
+    return pts, label
+
+
+train_dataset = tf.data.Dataset.from_tensor_slices((X_train, y_train))
+test_dataset = tf.data.Dataset.from_tensor_slices((X_test, y_test))
+
+train_dataset = train_dataset.shuffle(len(X_train)).map(augment).batch(BATCH_SIZE)
+test_dataset = test_dataset.shuffle(len(X_test)).batch(BATCH_SIZE)
+
+"""
+### Build a model
+"""
+
+"""
+Each convolution and fully-connected layer (with exception for end layers) consits of
+Convolution / Dense -> Batch Normalization -> ReLU Activation. We can make simple
+sub-classed layer to perform all these as a single layer.
+"""
+
+
+class ConvBN(layers.Layer):
+    def __init__(self, filters):
+        super(ConvBN, self).__init__()
+
+        self.conv = layers.Conv1D(filters, kernel_size=1, padding="valid")
+        self.bn = layers.BatchNormalization(momentum=0.0)
+        self.activation = tf.nn.relu
+
+    def call(self, inputs):
+
+        x = self.conv(inputs)
+        x = self.bn(x)
+        return self.activation(x)
+
+
+class DenseBN(layers.Layer):
+    def __init__(self, filters):
+        super(DenseBN, self).__init__()
+
+        self.dense = layers.Dense(filters)
+        self.bn = layers.BatchNormalization(momentum=0.0)
+        self.activation = tf.nn.relu
+
+    def call(self, inputs):
+
+        x = self.dense(inputs)
+        x = self.bn(x)
+        return self.activation(x)
+
+
+"""
+PointNet consists of two core components. The primary MLP network, and the transformer
+net (t-net). The t-net aims to learn an affine transformation matrix by its own mini
+network. The mini network strongly resembles the main network. The t-net is used twice.
+The first time to transform the input features ($n \times 3$) into a canonical
```

Comment:
Do not use LaTeX notation since we don't render it. Instead, just in-line Python code or pseudo code.